libs/vkd3d/command.c

```diff
@@ -8442,43 +8578,61 @@ static void STDMETHODCALLTYPE d3d12_command_list_ClearUnorderedAccessViewUint(d3
 
-        if (!uint_format)
+        if (uint_format)
+            vkd3d_mask_uint_clear_color(color.uint32, uint_format->vk_format);
+        else if (!vkd3d_clear_uav_uint_use_copy(base_view->format->dxgi_format))
         {
             ERR("Unhandled format %d.\n", base_view->format->dxgi_format);
             return;
         }
 
-        color = vkd3d_fixup_clear_uav_uint_color(list->device, base_view->format->dxgi_format, color);
-        vkd3d_mask_uint_clear_color(color.uint32, uint_format->vk_format);
-
         if (d3d12_resource_is_texture(resource_impl))
         {
-            struct vkd3d_texture_view_desc view_desc;
-            memset(&view_desc, 0, sizeof(view_desc));
+            if (uint_format)
+            {
+                struct vkd3d_texture_view_desc view_desc;
+                memset(&view_desc, 0, sizeof(view_desc));
+
+                view_desc.image = resource_impl->res.vk_image;
+                view_desc.view_type = base_view->info.texture.vk_view_type;
+                view_desc.format = uint_format;
+                view_desc.miplevel_idx = base_view->info.texture.miplevel_idx;
+                view_desc.miplevel_count = 1;
+                view_desc.layer_idx = base_view->info.texture.layer_idx;
+                view_desc.layer_count = base_view->info.texture.layer_count;
+                view_desc.aspect_mask = view_desc.format->vk_aspect_mask;
+                view_desc.image_usage = VK_IMAGE_USAGE_STORAGE_BIT;
+                view_desc.allowed_swizzle = false;
+
+                if (!vkd3d_create_texture_view(list->device, &view_desc, &args.u.view))
+                {
+                    ERR("Failed to create image view.\n");
+                    return;
+                }
 
-            view_desc.image = resource_impl->res.vk_image;
-            view_desc.view_type = base_view->info.texture.vk_view_type;
-            view_desc.format = uint_format;
-            view_desc.miplevel_idx = base_view->info.texture.miplevel_idx;
-            view_desc.miplevel_count = 1;
-            view_desc.layer_idx = base_view->info.texture.layer_idx;
-            view_desc.layer_count = base_view->info.texture.layer_count;
-            view_desc.aspect_mask = view_desc.format->vk_aspect_mask;
-            view_desc.image_usage = VK_IMAGE_USAGE_STORAGE_BIT;
-            view_desc.allowed_swizzle = false;
-
-            if (!vkd3d_create_texture_view(list->device, &view_desc, &args.u.view))
+                inline_view = args.u.view;
+            }
+            else
             {
-                ERR("Failed to create image view.\n");
-                return;
+                /* If the clear color is 0, we can safely use the existing view to perform the
+                 * clear since the bit pattern will not change. Otherwise, fill a scratch buffer
+                 * with the packed clear value and perform a buffer to image copy. */
+                if (color.uint32[0])
```

Comment:
If we have 0 color here and we hit the normal clear path, wouldn't B10G11R11 end up looking at the 3 float values instead, which might not have been cleared to zero?

Reply:
The fixup function zero-initializes its return value now.

Reply:
Right.